### DISCUSSION TITLE
defineClientAppSetup Usage & Vue 3 TS Cannot Find Module Type Declaration

### DISCUSSION THREAD

**Discussion by derogate** [vuepress/core#699] [Q&A] [answered]
Hey guys, new here. I have two questions. 

1. Regarding [`defineClientAppSetup` helper function](https://v2.vuepress.vuejs.org/reference/client-api.html#defineclientappsetup), it will be executed during `setup` hook, way before `beforeCreate` but is there anything useful or efficient that I can take advantage of this helper function during this stage?

2. In `.vuepress/clientAppEnhance.ts`, registering global components will give Typescript error `Cannot find module './components/<component-name>.vue' or its corresponding type declarations. ts(2307)` in the import component code. The component worked properly when used in *.md files though, just that the typescript error is a bit troubling. After some Googling & trial-and-error, I found out that I need to add a shim file `shims-vue.d.ts` with the content:

```
declare module '*.vue' {
	import type { DefineComponent } from 'vue'
	const component: DefineComponent<{}, {}, any>
	export default component
}
```

but it doesn't seem to resolve the Typescript issue when `shims.vue.data.ts` is not an actively opened file (when opened, the TS error is gone). Is it related to the bundler? Using the default vite without any bundleConfig btw

**Comment by Mister-Hope** [comment, marked as answer]
Fist of all, any vue using related issue are welcomed at vue forum or website like stackoverflow. And posing here is a bit out of topic.

I am not sure getting what you wanna ask in your first question. It will be called during client app setup, and what you want to do depends on yourslef. You can register some action using `onMounted` and you can watch route path to do something, also you can provide global computed variables. It's totally up to you to call everything in client app setup function.

And the second, I don't think vite will throw those issues, vite is using esbuild, and it does not have any static checking! If it's something your editor gives you, e.g. vscode, you will probably need a tsconfig.json to provide to your editor, so that they can read correct declaration files for their built-in typescript checker(or feature) to understand what `.vue` is.  (Also, that's not related to this project actually)

If you have other vuepress related question, they are always welcomed.🙂
